plutus-core/testlib/PlutusIR/Generators/QuickCheck/GenerateTerms.hs

```diff
@@ -0,0 +1,853 @@
+      pure (TyFun () a b, LamAbs () x a body)
+
+    genApp mty = withNoEscape $ do
+        ((_, arg), (toResTy, fun)) <-
+          astSizeSplit 1 4 (genTerm Nothing) (\ (argTy, _) -> genFun argTy mty)
+        case toResTy of
+          TyFun _ _ resTy -> pure (resTy, Apply () fun arg)
+          _               -> error $ display toResTy ++ "\n\n is not a 'TyFun'"
+      where
+        genFun argTy mty = genTerm . Just . TyFun () argTy =<< maybe (genType (Type ())) pure mty
+
+    genVarApp :: HasCallStack => _
+    genVarApp Nothing = withNoEscape $ do
+      -- CODE REVIEW: this function exists somewhere maybe? (Maybe even in this module...)
+      let arity (TyForall _ _ _ b) = 1 + arity b
+          arity (TyFun _ _ b)      = 1 + arity b
+          arity _                  = 0
+
+          appl :: HasCallStack => Int -> Term TyName Name DefaultUni DefaultFun () -> _
+          appl 0 tm b = return (b, tm)
+          appl n tm (TyForall _ x k b) = do
+            ty <- genType k
+            x' <- genLikelyFreshTyName "x"
+            appl (n - 1) (TyInst () tm ty) (substType (Map.singleton x' ty) $ renameVar x x' b)
+          appl n tm (TyFun _ a b) = do
+            (_, arg) <- genTerm (Just a)
+            appl (n - 1) (Apply () tm arg) b
+          appl _ _ _ = error "appl"
+
+          genV (x, ty0) = do
+            let ty = normalizeTy ty0
+            n <- liftGen $ choose (0, arity ty)
+            onAstSize (`div` n) $ appl n (Var () x) ty
+      asks (Map.toList . geTerms) >>= \ case
+        []   -> do
+          ty <- genType $ Type ()
+          (ty,) <$> inhabitType ty
+        vars -> oneof $ map genV vars
+
+    genVarApp (Just ty) = do
+      vars <- asks geTerms
+      ctx  <- asks geTypes
+      let cands = Map.toList vars
+          doInst _ tm (InstApp instTy) = pure $ TyInst () tm instTy
+          doInst n tm (InstArg argTy)  = onAstSize ((`div` n) . subtract 1)
+                                       . withNoEscape
+                                       $ Apply () tm <$> genTermOfType argTy
+      case [ foldM (doInst n) (Var () x) insts
+           | (x, a)      <- cands,
+             n           <- [0..typeArity a],
+             Right insts <- [findInstantiation ctx n ty a]
+           ] of
+        [] -> (ty,) <$> inhabitType ty
+        gs -> (ty,) <$> oneof gs
+
+-- | Like 'liftOf' except each of the elements is generated with a proportionally smaller size.
+scaledListOf :: GenTm a -> GenTm [a]
+scaledListOf g = do
+  sz <- asks geAstSize
+  n  <- choose (0, sz `div` 3)
+  onAstSize (`div` n) $ replicateM n g
+
+genDatatypeLet :: Bool -> (Datatype TyName Name DefaultUni () -> GenTm a) -> GenTm a
+genDatatypeLet rec cont = do
+    k <- liftGen arbitrary
+    let ks = argsFunKind k
+    n <- liftGen $ choose (1, 3)
+    ~(d : xs) <- genLikelyFreshTyNames $ "d" : replicate (length ks) "a"
+    ~(m : cs) <- genLikelyFreshNames   $ "m" : replicate n "c"
```

Comment:
We do, added this comment:

```
    -- Lazy matching to communicate to GHC the fact that this can't fail and thus doesn't require
    -- a 'MonadFail' (which 'GenTm' isn't).
```